frontend/src/app/workspace/service/copilot/texera-copilot.ts

```diff
@@ -0,0 +1,282 @@
+ * manipulating workflow components.
+ *
+ * State management includes:
+ * - UNAVAILABLE: Agent not initialized
+ * - AVAILABLE: Agent ready to receive messages
+ * - GENERATING: Agent currently processing and generating response
+ * - STOPPING: Agent in the process of stopping generation
+ */
+@Injectable()
+export class TexeraCopilot {
+  private model: any;
+  private modelType = "";
+  private agentName = "";
+  private messages: ModelMessage[] = [];
+  private agentResponses: AgentUIMessage[] = [];
+  private agentResponsesSubject = new BehaviorSubject<AgentUIMessage[]>([]);
+  public agentResponses$ = this.agentResponsesSubject.asObservable();
+  private state = CopilotState.UNAVAILABLE;
+  private stateSubject = new BehaviorSubject<CopilotState>(CopilotState.UNAVAILABLE);
+  public state$ = this.stateSubject.asObservable();
+
+  constructor(
+    private workflowActionService: WorkflowActionService,
+    private workflowUtilService: WorkflowUtilService,
+    private operatorMetadataService: OperatorMetadataService,
+    private workflowCompilingService: WorkflowCompilingService,
+    private notificationService: NotificationService
+  ) {}
+
+  public setAgentInfo(agentName: string): void {
+    this.agentName = agentName;
+  }
+
+  public setModelType(modelType: string): void {
+    this.modelType = modelType;
+  }
+
+  private setState(newState: CopilotState): void {
+    this.state = newState;
+    this.stateSubject.next(newState);
+  }
+  private emitAgentUIMessage(
+    role: "user" | "agent",
+    content: string,
+    isBegin: boolean,
+    isEnd: boolean,
+    toolCalls?: any[],
+    toolResults?: any[],
+    usage?: AgentUIMessage["usage"]
+  ): void {
+    this.agentResponses.push({ role, content, isBegin, isEnd, toolCalls, toolResults, usage });
+    this.agentResponsesSubject.next([...this.agentResponses]);
+  }
+  public initialize(): Observable<void> {
+    return defer(() => {
+      try {
+        this.model = createOpenAI({
+          baseURL: new URL(`${AppSettings.getApiEndpoint()}`, document.baseURI).toString(),
+          apiKey: "dummy",
+        }).chat(this.modelType);
+
+        this.setState(CopilotState.AVAILABLE);
+        return of(undefined);
+      } catch (error: unknown) {
+        this.setState(CopilotState.UNAVAILABLE);
+        return throwError(() => error);
+      }
+    });
+  }
+
+  public sendMessage(message: string): Observable<void> {
+    return defer(() => {
+      if (!this.model) {
+        return throwError(() => new Error("Copilot not initialized"));
+      }
+
+      if (this.state !== CopilotState.AVAILABLE) {
+        return throwError(() => new Error(`Cannot send message: agent is ${this.state}`));
+      }
+
+      this.setState(CopilotState.GENERATING);
+
+      this.emitAgentUIMessage("user", message, true, true);
+      this.messages.push({ role: "user", content: message });
+
+      const tools = this.createWorkflowTools();
+      let isFirstStep = true;
+
+      return from(
```

Comment:
This is the most import part of the life-cycle of agent handling. It would be good if you can have some docs here about what each step / callback does.